llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp

```diff
@@ -85,14 +85,16 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
 
   // Merge/Unmerge
   for (unsigned Op : {G_MERGE_VALUES, G_UNMERGE_VALUES}) {
-    unsigned BigTyIdx = Op == G_MERGE_VALUES ? 0 : 1;
-    unsigned LitTyIdx = Op == G_MERGE_VALUES ? 1 : 0;
     auto &MergeUnmergeActions = getActionDefinitionsBuilder(Op);
     if (XLen == 32 && ST.hasStdExtD()) {
-      LLT IdxZeroTy = G_MERGE_VALUES ? s64 : s32;
-      LLT IdxOneTy = G_MERGE_VALUES ? s32 : s64;
-      MergeUnmergeActions.legalFor({IdxZeroTy, IdxOneTy});
+      // Only need to legalize on IdxZeroTy since there is no way to construct
+      // a valid MERGE/UNMERGE where the types don't combine/decompose as
+      // expected.
+      LLT IdxZeroTy = Op == G_MERGE_VALUES ? s64 : s32;
+      MergeUnmergeActions.legalFor({IdxZeroTy});
```

Comment:
Don't you need to list 2 types here?

Reply:
Do we need to?
```
// Only need to legalize on IdxZeroTy since there is no way to construct
// a valid MERGE/UNMERGE where the types don't combine/decompose as
// expected.
```

Reply:
I'm surprised there isn't a check that the correct number of type indices is provided to .legalFor.

Reply:
I think there can only be two type indices for these opcodes. A small type or a large type. You cannot construct a MERGE/UNMERGE with >2 types.

Reply:
I'm surprised that we don't have to pass a 2 element initializer_list to `legalFor` for an opcode that is defined with 2 type indices.

Reply:
I have updated the code to do this, but I suppose there could be a case where you want to mark a type index as legal, regardless of what the other type index is.

Reply:
> I have updated the code to do this, but I suppose there could be a case where you want to mark a type index as legal, regardless of what the other type index is.

You couldn't make type index 1 legal and ignore type index 0 though. So it would be inconsistent if that was the intent.